Kube-router coredumping when cleaning up configuration

**What happened?**
We are unable to cleanup kube-router configuration like iptables etc.

**What did you expect to happen?**
We expected configuration changes made by kube-router to be reverted.

**How can we reproduce the behavior you experienced?**
Steps to reproduce the behavior:
```
kube-router --cleanup-config
```

** System Information (please complete the following information):**
 - Kube-Router Version (`kube-router --version`): 1.3.0 ( master branch as well )
 - Kube-Router Parameters: [e.g. --run-router --run-service-proxy --enable-overlay --overlay-type=full etc.]
 - Kubernetes Version (`kubectl version`) : 1.20
 - Kubernetes Deployment Type: K3s
 - Kube-Router Deployment Type: Systemd service
 - Cluster Size: 1 node

** Logs, other output, metrics **
```
❯ kube-router --cleanup-config
I0714 11:14:13.542887 2623569 network_policy_controller.go:525] Cleaning up iptables configuration permanently done by kube-router
E0714 11:14:13.578638 2623569 network_policy_controller.go:586] Failed to cleanup iptables rules: running [/usr/sbin/iptables -t filter -X KUBE-POD-FW-AH3EGJMMNT2MN6SA --wait]: exit status 1: iptables: Too many links.
I0714 11:14:13.578825 2623569 network_services_controller.go:2329] Cleaning up IPVS configuration permanently
E0714 11:14:13.589538 2623569 network_services_controller.go:638] Failed to run iptables command: running [/usr/sbin/iptables -t filter -D INPUT -m comment --comment handle traffic to IPVS service IPs in custom chain -m set --match-set kube-router-service-ips dst -j KUBE-ROUTER-SERVICES --wait]: exit status 2: iptables v1.8.7 (legacy): Couldn't load target `KUBE-ROUTER-SERVICES':No such file or directory
Try `iptables -h' or 'iptables --help' for more information.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17964a1]
goroutine 1 [running]:
sync.(*Mutex).Lock(...)
	/usr/lib/go-1.15/src/sync/mutex.go:74
github.com/cloudnativelabs/kube-router/pkg/controllers/proxy.(*NetworkServicesController).cleanupIpvsFirewall(0xc000957ae8)
	/root/go/src/github.com/cloudnativelabs/kube-router/pkg/controllers/proxy/network_services_controller.go:654 +0x1a1
github.com/cloudnativelabs/kube-router/pkg/controllers/proxy.(*NetworkServicesController).Cleanup(0xc000957ae8)
	/root/go/src/github.com/cloudnativelabs/kube-router/pkg/controllers/proxy/network_services_controller.go:2353 +0x2fc
github.com/cloudnativelabs/kube-router/pkg/cmd.CleanupConfigAndExit()
	/root/go/src/github.com/cloudnativelabs/kube-router/pkg/cmd/kube-router.go:65 +0xa5
main.Main(0xc000957f48, 0x449177)
	/root/go/src/github.com/cloudnativelabs/kube-router/cmd/kube-router/kube-router.go:63 +0x52a
main.main()
	/root/go/src/github.com/cloudnativelabs/kube-router/cmd/kube-router/kube-router.go:19 +0x26
```

--- a/pkg/controllers/netpol/network_policy_controller.go
+++ b/pkg/controllers/netpol/network_policy_controller.go
@@ -611,13 +611,18 @@ func (npc *NetworkPolicyController) Cleanup() {
 	}
 
 	// delete all ipsets
-	klog.V(1).Infof("Attempting to attain ipset mutex lock")
-	npc.ipsetMutex.Lock()
-	klog.V(1).Infof("Attained ipset mutex lock, continuing...")
-	defer func() {
-		npc.ipsetMutex.Unlock()
-		klog.V(1).Infof("Returned ipset mutex lock")
-	}()
+	// There are certain actions like Cleanup() actions that aren't working with full instantiations of the controller
+	// and in these instances the mutex may not be present and may not need to be present as they are operating out of a
+	// single goroutine where there is no need for locking
+	if nil != npc.ipsetMutex {
+		klog.V(1).Infof("Attempting to attain ipset mutex lock")
+		npc.ipsetMutex.Lock()
+		klog.V(1).Infof("Attained ipset mutex lock, continuing...")
+		defer func() {
+			npc.ipsetMutex.Unlock()
+			klog.V(1).Infof("Returned ipset mutex lock")
+		}()
+	}
 	ipset, err := utils.NewIPSet(false)
 	if err != nil {
 		klog.Errorf("Failed to clean up ipsets: " + err.Error())

--- a/pkg/controllers/proxy/network_services_controller.go
+++ b/pkg/controllers/proxy/network_services_controller.go
@@ -650,13 +650,18 @@ func (nsc *NetworkServicesController) cleanupIpvsFirewall() {
 	}
 
 	// Clear ipsets.
-	klog.V(1).Infof("Attempting to attain ipset mutex lock")
-	nsc.ipsetMutex.Lock()
-	klog.V(1).Infof("Attained ipset mutex lock, continuing...")
-	defer func() {
-		nsc.ipsetMutex.Unlock()
-		klog.V(1).Infof("Returned ipset mutex lock")
-	}()
+	// There are certain actions like Cleanup() actions that aren't working with full instantiations of the controller
+	// and in these instances the mutex may not be present and may not need to be present as they are operating out of a
+	// single goroutine where there is no need for locking
+	if nil != nsc.ipsetMutex {
+		klog.V(1).Infof("Attempting to attain ipset mutex lock")
+		nsc.ipsetMutex.Lock()
+		klog.V(1).Infof("Attained ipset mutex lock, continuing...")
+		defer func() {
+			nsc.ipsetMutex.Unlock()
+			klog.V(1).Infof("Returned ipset mutex lock")
+		}()
+	}
 	ipSetHandler, err := utils.NewIPSet(false)
 	if err != nil {
 		klog.Errorf("Failed to initialize ipset handler: %s", err.Error())

--- a/pkg/controllers/routing/network_routes_controller.go
+++ b/pkg/controllers/routing/network_routes_controller.go
@@ -646,13 +646,18 @@ func (nrc *NetworkRoutingController) Cleanup() {
 	}
 
 	// delete all ipsets created by kube-router
-	klog.V(1).Infof("Attempting to attain ipset mutex lock")
-	nrc.ipsetMutex.Lock()
-	klog.V(1).Infof("Attained ipset mutex lock, continuing...")
-	defer func() {
-		nrc.ipsetMutex.Unlock()
-		klog.V(1).Infof("Returned ipset mutex lock")
-	}()
+	// There are certain actions like Cleanup() actions that aren't working with full instantiations of the controller
+	// and in these instances the mutex may not be present and may not need to be present as they are operating out of a
+	// single goroutine where there is no need for locking
+	if nil != nrc.ipsetMutex {
+		klog.V(1).Infof("Attempting to attain ipset mutex lock")
+		nrc.ipsetMutex.Lock()
+		klog.V(1).Infof("Attained ipset mutex lock, continuing...")
+		defer func() {
+			nrc.ipsetMutex.Unlock()
+			klog.V(1).Infof("Returned ipset mutex lock")
+		}()
+	}
 	ipset, err := utils.NewIPSet(nrc.isIpv6)
 	if err != nil {
 		klog.Errorf("Failed to clean up ipsets: " + err.Error())

